<a href="https://colab.research.google.com/github/Shreyy1203/Hiring-Prediction-Model/blob/main/Hiring.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Name : Shreyash Shinde  
# Email : connect.shindeshreyash@gmail.com  

# Task 1 -Load and Understand the Dataset

### Step 1 - Import Libraries

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

### Step 2 - Load the Dataset & Display Data Samples


In [ ]:
data = pd.read_csv("Resume_List.csv")
data.head(5)

,Resume_ID,Name,Skills,Experience,Education,Certifications,Job_Role,Recruiter Decision,Salary,Projects,AI Score (0-100)
0,1,Ashley Ali,"TensorFlow, NLP, Pytorch",10,B.Sc,NaN,AI Researcher,Hire,104895,8,100
1,2,Wesley Roman,"Deep Learning, Machine Learning, Python, SQL",10,MBA,Google ML,Data Scientist,Hire,113002,1,100
2,3,Corey Sanchez,"Ethical Hacking, Cybersecurity, Linux",1,MBA,Deep Learning Specialization,Cybersecurity Analyst,Hire,71766,7,70
3,4,Elizabeth Carney,"Python, Pytorch, TensorFlow",7,B.Tech,AWS Certified,AI Researcher,Hire,46848,0,95
4,5,Julie Hill,"SQL, React, Java",4,PhD,NaN,Software Engineer,Hire,87441,9,100


In [ ]:
data.tail(5)

,Resume_ID,Name,Skills,Experience,Education,Certifications,Job_Role,Recruiter Decision,Salary,Projects,AI Score (0-100)
995,996,Brenda Williams,"Cybersecurity, Linux, Ethical Hacking",0,B.Sc,NaN,Cybersecurity Analyst,Reject,114364,9,60
996,997,Colleen Hicks,"Deep Learning, Machine Learning",0,MBA,Deep Learning Specialization,Data Scientist,Reject,103294,5,45
997,998,Michelle Molina,"TensorFlow, NLP",0,B.Tech,Google ML,AI Researcher,Hire,113855,9,65
998,999,Danielle Horn,"Linux, Networking, Cybersecurity, Ethical Hacking",8,PhD,AWS Certified,Cybersecurity Analyst,Hire,83146,10,100
999,1000,Chad Collins,"SQL, Machine Learning, Python, Deep Learning",7,M.Tech,Deep Learning Specialization,Data Scientist,Hire,119474,3,100


In [ ]:
data.sample(5)

,Resume_ID,Name,Skills,Experience,Education,Certifications,Job_Role,Recruiter Decision,Salary,Projects,AI Score (0-100)
736,737,Matthew Reed,"Ethical Hacking, Linux, Networking",3,M.Tech,NaN,Cybersecurity Analyst,Reject,95288,3,60
762,763,Mr. Johnathan Garcia,"SQL, Java, React",2,B.Tech,NaN,Software Engineer,Reject,82150,4,55
403,404,Gregg Barber,"Python, Machine Learning, Deep Learning",0,M.Tech,Deep Learning Specialization,Data Scientist,Reject,114745,6,55
475,476,Christina Oneal,"Python, Machine Learning",10,MBA,Google ML,Data Scientist,Hire,89482,5,100
874,875,Wayne Hudson,"C++, React, SQL, Java",5,B.Sc,AWS Certified,Software Engineer,Hire,84438,10,100


### Step 3 - Understanding the Data

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Resume_ID           1000 non-null   int64 
 1   Name                1000 non-null   object
 2   Skills              1000 non-null   object
 3   Experience          1000 non-null   int64 
 4   Education           1000 non-null   object
 5   Certifications      726 non-null    object
 6   Job_Role            1000 non-null   object
 7   Recruiter Decision  1000 non-null   object
 8   Salary              1000 non-null   int64 
 9   Projects            1000 non-null   int64 
 10  AI Score (0-100)    1000 non-null   int64 
dtypes: int64(5), object(6)
memory usage: 86.1+ KB


### Short Explaination of the data  
Q. What type of data is present?  
=> There are 2 different types of data present in the dataset viz., 5 Numerical (int64) and 6 Categorical (object).  
Numerical Datas are Resume_ID, Experience, Salary Expectation, Project Counts and AI Score.  
Categorical Datas are Name, Skills, Education, Certifications, Job Role, Recruiter Descision.

Q. Which Column would be used as target?  
The Recruiter Decision column would be used as the target.

# Task 2 - Basic Data Inspection

In [ ]:
print ("Number of Rows: ", data.shape[0])
print ("Number of Columns: ", data.shape[1])
print ("Column Names: ", data.columns)
print ("Data Types:\n", data.dtypes)
print (data['Recruiter Decision'].value_counts())
print ("Summary :")
data.describe()

Number of Rows:  1000
Number of Columns:  11
Column Names:  Index(['Resume_ID', 'Name', 'Skills', 'Experience', 'Education',
       'Certifications', 'Job_Role', 'Recruiter Decision', 'Salary',
       'Projects', 'AI Score (0-100)'],
      dtype='object')
Data Types:
 Resume_ID              int64
Name                  object
Skills                object
Experience             int64
Education             object
Certifications        object
Job_Role              object
Recruiter Decision    object
Salary                 int64
Projects               int64
AI Score (0-100)       int64
dtype: object
Recruiter Decision
Hire      812
Reject    188
Name: count, dtype: int64
Summary :


,Resume_ID,Experience,Salary,Projects,AI Score (0-100)
count,1000.000000,1000.000000,1000.000000,1000.00000,1000.000000
mean,500.500000,4.896000,79994.486000,5.13300,83.950000
std,288.819436,3.112695,23048.472549,3.23137,20.983036
min,1.000000,0.000000,40085.000000,0.00000,15.000000
25%,250.750000,2.000000,60415.750000,2.00000,70.000000
50%,500.500000,5.000000,79834.500000,5.00000,100.000000
75%,750.250000,8.000000,99583.250000,8.00000,100.000000
max,1000.000000,10.000000,119901.000000,10.00000,100.000000


Before training a model, data inspection is critical for several reasons:  
**Detecting Imbalance:** If 90% of the data is "Hire," the model might just learn to say "Hire" every time to be 90% accurate without actually learning the skills. We check value_counts to see if we need to balance the data.  
**Identifying Data Types:** Machine Learning models are essentially giant calculators—they cannot "read" text. Inspection tells us which columns need to be converted into numbers (Encoding).  
**Spotting Errors/Outliers:** If the Salary Expectation ($) has a minimum value of 0 or a maximum of 10 million, it could skew the "average" and confuse the AI.    
**Feature Selection:** It helps us confirm which columns are actually useful. For instance, seeing that Resume_ID is just a sequence of numbers (1, 2, 3...) confirms it has no predictive power and should be dropped.

# Task 3 - Data Cleaning

### Step 1 - Drop the useless columns

In [ ]:
data_clean = data.drop(columns=['Resume_ID', 'Name'])
data_clean.sample(5)

,Skills,Experience,Education,Certifications,Job_Role,Recruiter Decision,Salary,Projects,AI Score (0-100)
240,"Linux, Ethical Hacking",10,MBA,Deep Learning Specialization,Cybersecurity Analyst,Hire,50784,0,100
258,"SQL, Machine Learning, Deep Learning",0,B.Tech,Deep Learning Specialization,Data Scientist,Reject,113435,2,35
954,"SQL, Python, Machine Learning, Deep Learning",7,B.Sc,AWS Certified,Data Scientist,Hire,55431,2,100
473,"SQL, Machine Learning, Deep Learning",9,B.Tech,NaN,Data Scientist,Hire,114686,0,100
650,"Python, Pytorch, NLP, TensorFlow",4,PhD,Deep Learning Specialization,AI Researcher,Hire,50369,8,100


### Step 2 - Converting Recruiter Descision into Binary

In [ ]:
data_clean['Recruiter Decision'] = data_clean['Recruiter Decision'].replace({'Hire': 1, 'Reject' : 0})
data_clean.sample(5)

/tmp/ipykernel_478/1854397919.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data_clean['Recruiter Decision'] = data_clean['Recruiter Decision'].replace({'Hire': 1, 'Reject' : 0})


,Skills,Experience,Education,Certifications,Job_Role,Recruiter Decision,Salary,Projects,AI Score (0-100)
468,"C++, Java, SQL",2,MBA,Google ML,Software Engineer,1,62255,7,80
422,"NLP, TensorFlow, Pytorch, Python",8,B.Sc,Deep Learning Specialization,AI Researcher,1,108667,0,100
365,"TensorFlow, Pytorch",5,M.Tech,Deep Learning Specialization,AI Researcher,1,82093,9,100
326,"Machine Learning, Python, Deep Learning",9,B.Tech,NaN,Data Scientist,1,93707,0,100
626,"SQL, C++",0,B.Tech,Google ML,Software Engineer,0,46679,5,45


### Step 3 - Checking Missing Values

In [ ]:
print (data_clean.isnull().sum())

Skills                  0
Experience              0
Education               0
Certifications        274
Job_Role                0
Recruiter Decision      0
Salary                  0
Projects                0
AI Score (0-100)        0
dtype: int64


There are 274 missing values in certification column. We will not drop them, as it will delete 27.4% of our data. Instead we will convert them into 'none'

In [ ]:
data_clean['Certifications'] = data_clean['Certifications'].fillna('none')
print (data_clean.isnull().sum())

Skills                0
Experience            0
Education             0
Certifications        0
Job_Role              0
Recruiter Decision    0
Salary                0
Projects              0
AI Score (0-100)      0
dtype: int64


### Step 4 - Ensure numeric columns are corectly formatted

In [ ]:
print (data_clean.dtypes)

Skills                object
Experience             int64
Education             object
Certifications        object
Job_Role              object
Recruiter Decision     int64
Salary                 int64
Projects               int64
AI Score (0-100)       int64
dtype: object


As we can see, the columns 'Experience', 'Recruiter Descision', 'Salary Expectation', 'Project Count' and 'AI Score' all are integer values.

# Task 4 - Text Feature Engineering

### step 1 - Combine text columns into one

In [ ]:
import re
data_clean['combined_text'] = data_clean['Skills'] + ' ' + data_clean['Certifications'] + ' ' + data_clean['Job_Role']
data_clean['combined_text']

,combined_text
0,"TensorFlow, NLP, Pytorch none AI Researcher"
1,"Deep Learning, Machine Learning, Python, SQL G..."
2,"Ethical Hacking, Cybersecurity, Linux Deep Lea..."
3,"Python, Pytorch, TensorFlow AWS Certified AI R..."
4,"SQL, React, Java none Software Engineer"
...,...
995,"Cybersecurity, Linux, Ethical Hacking none Cyb..."
996,"Deep Learning, Machine Learning Deep Learning ..."
997,"TensorFlow, NLP Google ML AI Researcher"
998,"Linux, Networking, Cybersecurity, Ethical Hack..."


# Step 2 - Convert all into Lowercase

In [ ]:
data_clean['combined_text'] = data_clean['combined_text'].str.lower()
data_clean['combined_text']

,combined_text
0,"tensorflow, nlp, pytorch none ai researcher"
1,"deep learning, machine learning, python, sql g..."
2,"ethical hacking, cybersecurity, linux deep lea..."
3,"python, pytorch, tensorflow aws certified ai r..."
4,"sql, react, java none software engineer"
...,...
995,"cybersecurity, linux, ethical hacking none cyb..."
996,"deep learning, machine learning deep learning ..."
997,"tensorflow, nlp google ml ai researcher"
998,"linux, networking, cybersecurity, ethical hack..."


### Step 3 - Remove whitespaces

In [ ]:
data_clean['combined_text'] = data_clean['combined_text'].str.replace(r'[^a-z0-9\s]', ' ', regex=True)
data_clean['combined_text']

,combined_text
0,tensorflow nlp pytorch none ai researcher
1,deep learning machine learning python sql g...
2,ethical hacking cybersecurity linux deep lea...
3,python pytorch tensorflow aws certified ai r...
4,sql react java none software engineer
...,...
995,cybersecurity linux ethical hacking none cyb...
996,deep learning machine learning deep learning ...
997,tensorflow nlp google ml ai researcher
998,linux networking cybersecurity ethical hack...


### Step 4 - Remove Extra Spaces

In [ ]:
data_clean['combined_text'] = data_clean['combined_text'].str.replace(r'\s+', ' ', regex=True).str.strip()
data_clean['combined_text']

,combined_text
0,tensorflow nlp pytorch none ai researcher
1,deep learning machine learning python sql goog...
2,ethical hacking cybersecurity linux deep learn...
3,python pytorch tensorflow aws certified ai res...
4,sql react java none software engineer
...,...
995,cybersecurity linux ethical hacking none cyber...
996,deep learning machine learning deep learning s...
997,tensorflow nlp google ml ai researcher
998,linux networking cybersecurity ethical hacking...


Lets compare original with combined

In [ ]:
data_clean[['Skills', 'Certifications', 'Job_Role', 'combined_text']]

,Skills,Certifications,Job_Role,combined_text
0,"TensorFlow, NLP, Pytorch",none,AI Researcher,tensorflow nlp pytorch none ai researcher
1,"Deep Learning, Machine Learning, Python, SQL",Google ML,Data Scientist,deep learning machine learning python sql goog...
2,"Ethical Hacking, Cybersecurity, Linux",Deep Learning Specialization,Cybersecurity Analyst,ethical hacking cybersecurity linux deep learn...
3,"Python, Pytorch, TensorFlow",AWS Certified,AI Researcher,python pytorch tensorflow aws certified ai res...
4,"SQL, React, Java",none,Software Engineer,sql react java none software engineer
...,...,...,...,...
995,"Cybersecurity, Linux, Ethical Hacking",none,Cybersecurity Analyst,cybersecurity linux ethical hacking none cyber...
996,"Deep Learning, Machine Learning",Deep Learning Specialization,Data Scientist,deep learning machine learning deep learning s...
997,"TensorFlow, NLP",Google ML,AI Researcher,tensorflow nlp google ml ai researcher
998,"Linux, Networking, Cybersecurity, Ethical Hacking",AWS Certified,Cybersecurity Analyst,linux networking cybersecurity ethical hacking...


# Task 5 - Convert text into numerical features

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(data_clean['combined_text'])
tfidf_data = pd.DataFrame(tfidf_matrix.toarray(), columns=tfidf.get_feature_names_out())
data_final = pd.concat([data_clean.drop(columns=['Skills', 'Certifications', 'Job_Role', 'combined_text']), tfidf_data], axis=1)
print (f"Number of Unique Text Features Extracted : {tfidf_matrix.shape[1]}")

Number of Unique Text Features Extracted : 28


Q. Why ML models cannot work directly with text?  
=> Machine Learning models (like Random Forests or Logistic Regression) are essentially giant mathematical formulas (e.g., y=mx+b). You can multiply a salary of $100,000 by a weight of 0.5, but you cannot mathematically multiply the string "Python" by 0.5. Text must be converted to numbers so the algorithm can compute distance, probabilities, and patterns.

Q. What TF-IDF represents?
TF-IDF stands for Term Frequency-Inverse Document Frequency. It is a scoring system used to determine how important a specific word is to a resume.  
TF (Term Frequency): How often a word appears in a specific candidate's resume.  
IDF (Inverse Document Frequency): How rare the word is across all 1,000 resumes.

# Task 6 - Encode Categorical Variables

In [ ]:
le = LabelEncoder()
data_final['Education_Encoded'] = le.fit_transform(data_final['Education'])
data_final = data_final.drop(columns=['Education'])
data_final.sample(5)

,Experience,Recruiter Decision,Salary,Projects,AI Score (0-100),ai,analyst,aws,certified,cybersecurity,...,python,pytorch,react,researcher,scientist,software,specialization,sql,tensorflow,Education_Encoded
37,5,1,97505,7,100,0.000000,0.0,0.000000,0.000000,0.0,...,0.0,0.000000,0.000000,0.000000,0.0,0.410979,0.0,0.333776,0.0,1
104,5,1,81801,1,75,0.000000,0.0,0.000000,0.000000,0.0,...,0.0,0.000000,0.000000,0.000000,0.0,0.410979,0.0,0.333776,0.0,1
563,6,1,103874,5,100,0.000000,0.0,0.385389,0.385389,0.0,...,0.0,0.000000,0.451492,0.000000,0.0,0.396066,0.0,0.000000,0.0,0
880,9,1,50001,6,100,0.000000,0.0,0.402212,0.402212,0.0,...,0.0,0.000000,0.471200,0.000000,0.0,0.413355,0.0,0.335706,0.0,1
623,5,1,64489,4,90,0.389918,0.0,0.395132,0.395132,0.0,...,0.0,0.440555,0.000000,0.389918,0.0,0.000000,0.0,0.000000,0.0,2


Q. Explain the difference between Label Enoding and One Hot Encoding.
=> Label Encoding:

1.   Label Encoding
*   How it works: It replaces each text category with a single integer (0, 1, 2, 3...).
*   When to use it: When there is an ordinal relationship (a natural rank or order). For example, Education is a great fit for this because a PhD is technically a "higher" degree level than a B.Sc.
* The Risk: Some ML models might incorrectly assume that an MBA (3) is "three times better" than a B.Tech (1) because 3>1.


2. One-Hot Encoding:
* How it works: It creates a brand new binary column (1 or 0) for every single category. Instead of one column called "Education", you get 5 columns: Is_BSc, Is_BTech, Is_MTech, Is_MBA, and Is_PhD. If a candidate has a PhD, their Is_PhD column equals 1, and the others equal 0.
* When to use it: When there is a nominal relationship (no rank or order). For example, if you had a column for "Favorite Color" (Red, Blue, Green), one color isn't mathematically higher than the other.
* The Risk: It creates the "Curse of Dimensionality." If you have 100 different categories, you suddenly add 100 new columns to your dataset, making your model much slower and prone to overfitting.

# Task 7 - Feature and Target Sepration

In [ ]:
X = data_final.drop (columns=['Recruiter Decision', 'AI Score (0-100)'])
y = data_final['Recruiter Decision']


Q. Why should Recruiter Decision NOT be included in X?  
=>If you leave the Recruiter Decision column inside your features (X), you will encounter a massive problem called Data Leakage.  
Think of it like giving a student an exam where the answer key is printed right next to the questions.  
The Model Memorizes Instead of Learns: The algorithm will look at the data and immediately realize, "Oh, if the 'Recruiter Decision' column says 1, the answer is 1!"  
100% Accuracy Illusion: During testing, the model will score 100% accuracy because it is cheating.  
Real-World Failure: When HR actually tries to use this system on a brand new, unseen resume, there won't be a "Recruiter Decision" column yet (because that's what the AI is supposed to figure out). The model will immediately crash or make random guesses because its "answer key" is missing.  
By separating X and y, we force the AI to genuinely learn the mathematical relationship between a candidate's skills, experience, and education (X) and the final hiring outcome (y).

# Task 8 - Spliting the dataset

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print ("Training Data: ", X_train.shape, y_train.shape)
print ("Testing Data: ", X_test.shape, y_test.shape)

Training Data:  (800, 32) (800,)
Testing Data:  (200, 32) (200,)


Q. Why is a Train-Test Split Necessary?   
=> Imagine you are a professor teaching a math class. You give your students a study guide with 100 practice problems (the Training Data).  
When it is time for the final exam, would you give them the exact same 100 problems from the study guide? No! If you did, you wouldn't know if they actually learned the math concepts or if they just memorized the answers. You need to give them 20 brand-new problems (the Testing Data) to see how well they perform in the real world.
In Machine Learning, if we train the model on 100% of our data, we have no way to test if it's actually good at predicting new resumes. By holding back 20% of the data as a "secret test," we can evaluate the model's true accuracy.

Q. What Does "Overfitting" Mean?  
=> Overfitting happens when a Machine Learning model learns the training data too perfectly.  
Going back to the student analogy: an overfitted model is the student who memorized that "Question 4's answer is C" instead of learning the actual formula.
If your model overfits the HR dataset, it might memorize hyper-specific, useless patterns—for example, it might learn that "if a candidate has exactly 3 projects, 4 years of experience, and their name is on row 42, hire them."

# Task 9 - Feature Scaling

In [ ]:
scaler = StandardScaler()
cols_to_scale = ['Experience', 'Salary', 'Projects']

print ("Before Scaling")
print (X_train[cols_to_scale].head(3))
print (X_test[cols_to_scale].head(3))

X_train_scaled = X_train.copy()
X_test_scaled = X_test.copy()

X_train_scaled[cols_to_scale] = scaler.fit_transform(X_train[cols_to_scale])
X_test_scaled[cols_to_scale] = scaler.transform(X_test[cols_to_scale])

print ("After Scaling")
print(X_train_scaled[cols_to_scale].head(3))
print (X_test_scaled[cols_to_scale].head(3))

Before Scaling
     Experience  Salary  Projects
29           10   59926         4
535           4  108770        10
695           3  118062         8
     Experience  Salary  Projects
521           9  108379        10
737           6   56096         8
740           5   77117         6
After Scaling
     Experience    Salary  Projects
29     1.639910 -0.894401 -0.387502
535   -0.297945  1.259795  1.489401
695   -0.620921  1.669606  0.863766
     Experience    Salary  Projects
521    1.316934  1.242551  1.489401
737    0.348007 -1.063318  0.863766
740    0.025031 -0.136216  0.238132


Q. Why is Scaling Important?  
=>Machine Learning algorithms use raw math. Because Salary of 100,000 is infinitely larger than Experience of 5, an unscaled model will incorrectly assume that Salary is the only thing that matters and completely ignore the candidate's Experience and Projects.  
StandardScaler fixes this by transforming the data so that every numerical column has a mean (average) of 0 and a standard deviation of 1. Now, all features share the exact same mathematical weight, allowing the AI to evaluate them fairly.

Q. Which Models Require Scaling?  
* Models that DO require scaling:  
Distance-Based Models: K-Nearest Neighbors (KNN), Support Vector Machines (SVM). These calculate the physical geometric distance between data points; if salary is in the thousands, the distance calculation explodes.
Gradient Descent-Based Models: Logistic Regression, Neural Networks, Linear Regression. These use calculus to find the best fit. Huge numbers slow down the math and stop the model from learning properly.
* Models that DO NOT require scaling:  
Tree-Based Models: Decision Trees, Random Forests, XGBoost. These models don't calculate distances or gradients. They just ask "Yes/No" questions, like "Is Salary > 80,000?". Scaling doesn't affect how a tree makes a split.

# Task 10 - Model Training

Models which use scaled data : Logistic Regression, Support Vector Machine, K-Nearest Neighbours.

Models which do not use scaled data : Random Forest Classifier.

In [ ]:
#Logistic Regression
lr_model = LogisticRegression()
lr_model.fit(X_train_scaled, y_train)
lr_predictions = lr_model.predict(X_test_scaled)

#Support Vector Machine
svm_model = SVC()
svm_model.fit(X_train_scaled, y_train)
svm_predictions = svm_model.predict(X_test_scaled)

#K-Nearest Neighbours
knn_model = KNeighborsClassifier()
knn_model.fit(X_train_scaled, y_train)
knn_predictions = knn_model.predict(X_test_scaled)

#Random Forest Classifier
rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)
rf_predictions = rf_model.predict(X_test)

#Accuracies
print ("Logistic Regression Accuracy: ", accuracy_score(y_test, lr_predictions))
print ("Support Vector Machine Accuracy: ", accuracy_score(y_test, svm_predictions))
print ("K-Nearest Neighbours Accuracy: ", accuracy_score(y_test, knn_predictions))
print ("Random Forest Classifier Accuracy: ", accuracy_score(y_test, rf_predictions))

Logistic Regression Accuracy:  0.98
Support Vector Machine Accuracy:  0.98
K-Nearest Neighbours Accuracy:  0.945
Random Forest Classifier Accuracy:  0.95


Q. How each model works?
* Logistic Regression:  
Despite the word "regression," this is a classification tool. It takes all the candidate's features, multiplies them by learned weights, and passes the result through an "S-shaped" curve (the Sigmoid function). This outputs a probability between 0 and 1. If the probability is above 0.5 (50%), the model classifies the candidate as a Hire.

* Support Vector Machine:   
Imagine plotting all your candidates on a multi-dimensional graph, where Hires are green dots and Rejects are red dots. SVM attempts to draw a physical boundary line (called a hyperplane) directly between the two groups. It specifically tries to maximize the empty space (the margin) between the line and the closest data points to ensure the classification is as robust as possible.

* K-Nearest Neighbours (KNN):  
This is the most intuitive model. It essentially maps out all the candidates. When a new resume comes in, KNN looks at the "K" number of closest candidates in the dataset (e.g., the 5 candidates with the most similar experience, salary, and skills). If the majority of those 5 closest neighbors were Hired, the new candidate is also classified as a Hire.

* Random Forest Classifier:  
This is an "ensemble" model. Instead of relying on one decision process, it builds hundreds of individual Decision Trees during training. Each tree asks a series of Yes/No questions (e.g., "Is Experience > 5?", "Does the resume contain the word 'Python'?"). When a new resume is evaluated, all the trees make a prediction, and the forest takes a majority vote to determine the final decision.

# Task 11 - Model Evaluation and Comparision

In [ ]:
#Accuracies
print ("Logistic Regression Accuracy: ", accuracy_score(y_test, lr_predictions))
print ("Support Vector Machine Accuracy: ", accuracy_score(y_test, svm_predictions))
print ("K-Nearest Neighbours Accuracy: ", accuracy_score(y_test, knn_predictions))
print ("Random Forest Classifier Accuracy: ", accuracy_score(y_test, rf_predictions))

#Classification Report
print('Logistic Regression Report')
print(classification_report(y_test, lr_predictions))

print ('Support Vector Machine Report')
print(classification_report(y_test, svm_predictions))

print ('K-Nearest Neighbours Report')
print(classification_report(y_test, knn_predictions))

print ('Random Forest Classifier Report')
print(classification_report(y_test, rf_predictions))

result = pd.DataFrame ({
    'Model' : ['Logistic Regression', 'Support Vector Machine', 'K-Nearest Neighbours', 'Random Forest Classifier'],
    'Accuracy' : [accuracy_score(y_test, lr_predictions), accuracy_score(y_test, svm_predictions), accuracy_score(y_test, knn_predictions), accuracy_score(y_test, rf_predictions)]
})
result.sort_values(by='Accuracy', ascending=False)

Logistic Regression Accuracy:  0.98
Support Vector Machine Accuracy:  0.98
K-Nearest Neighbours Accuracy:  0.945
Random Forest Classifier Accuracy:  0.95
Logistic Regression Report
              precision    recall  f1-score   support

           0       0.94      0.98      0.96        46
           1       0.99      0.98      0.99       154

    accuracy                           0.98       200
   macro avg       0.97      0.98      0.97       200
weighted avg       0.98      0.98      0.98       200

Support Vector Machine Report
              precision    recall  f1-score   support

           0       0.94      0.98      0.96        46
           1       0.99      0.98      0.99       154

    accuracy                           0.98       200
   macro avg       0.97      0.98      0.97       200
weighted avg       0.98      0.98      0.98       200

K-Nearest Neighbours Report
              precision    recall  f1-score   support

           0       0.89      0.87      0.88        4

,Model,Accuracy
0,Logistic Regression,0.980
1,Support Vector Machine,0.980
3,Random Forest Classifier,0.950
2,K-Nearest Neighbours,0.945


# Task 13 - Build Hiring Prediction Function

Building a function that accepts user input resume and predicts the result.

In [ ]:
4def resume (Skills, Experience, Education, Certifications, Job_Role, Salary, Projects):
  #Clean and combine text
  combined_text = f"{Skills} {Certifications} {Job_Role}".lower()
  combined_text = re.sub(r'[^a-z0-9\s]', ' ', combined_text)
  combined_text = re.sub(r'\s+', ' ', combined_text).strip()

  #TF-IDF Transformation
  text_features = tfidf.transform([combined_text]).toarray()

  #Label Encoding
  try :
    encoded = le.transform([Education])[0]
  except ValueError :
    encoded = le.transform(['B.Sc'])[0]

  numeric_features = scaler.transform([[Experience, Salary, Projects]])

  resume_final = np.hstack ((numeric_features, [[encoded]], text_features))

  prediction = lr_model.predict(resume_final)
  probability = lr_model.predict_proba(resume_final)

  decision = 'Hire' if prediction[0] == 1 else 'Reject'
  confidence = probability[0][1] if prediction == 1 else probability[0][0]

  print ("Decision: ", decision)
  print ("Confidence: ", confidence * 100)

  return decision, confidence

print ('Testing')
Skills = input ('Enter Skills seperated by commas : ')
Experience = int (input ('Enter Experience : '))
Education = input ('Enter Education : ')
Certifications = input ('Enter Certifications : ')
Job_Role = input ('Enter Job Role : ')
Salary = int (input ('Enter Salary : '))
Projects = int (input ('Enter Projects : '))

resume (Skills, Experience, Education, Certifications, Job_Role, Salary, Projects)

Testing
Enter Skills seperated by commas : Java, Python
Enter Experience : 1
Enter Education : B. Tech
Enter Certifications : None
Enter Job Role : Data Scientist
Enter Salary : 100000
Enter Projects : 2
Decision:  Reject
Confidence:  96.76067328228538


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


('Reject', np.float64(0.9676067328228538))

# Task 14 - Final Conclusion

1. Dataset Understanding:  
The project utilized a dataset consisting of 1,000 synthetic candidate resumes. The data featured a mix of continuous numerical variables (e.g., Experience, Salary Expectations, Projects) and unstructured text data (e.g., Skills, Education, Certifications). The core objective was to solve a supervised binary classification problem: accurately predicting the Recruiter Decision (mapped to 1 for Hire, 0 for Reject).
---
2. Key Preprossesing Steps:  
Transforming raw human language and varied data points into a machine-readable format was the most critical phase of the project. Key steps included:  
* Data Cleaning: Removing non-predictive identifiers (Resume_ID, Name) and handling missing values to prevent data loss.
* Text Engineering (NLP): Combining textual features, cleaning special characters, and applying TF-IDF Vectorization to convert words into weighted numerical matrices based on their frequency and importance.
* Encoding & Scaling: Applying Label Encoding to ordinal categorical data (Education) and using StandardScaler to normalize numerical features. This ensured that large-scale numbers (like salary) didn't mathematically overpower smaller-scale numbers (like years of experience).
---
3. Best Performing Model  
After evaluating four distinct algorithms (Logistic Regression, Random Forest, SVM, and KNN), Logistic Regression and Support Vector Machine (SVM) emerged as the top performers, both achieving 98.00% accuracy. After hyperparameter tuning with GridSearchCV, Logistic Regression (or a linear SVM) was proven to be the most practical choice. It is computationally efficient and outputs clear probability scores, making it ideal for datasets with clear, linear relationships between qualifications and hiring decisions.
---
4. What you learned  
This project reinforced that machine learning is 80% data preparation and 20% algorithm tuning. Key technical takeaways include:
* The absolute necessity of train-test splits to prevent model overfitting and to accurately evaluate how the system will perform on unseen data.
* The power of scikit-learn Pipelines. Bundling preprocessing steps (scaling, vectorizing) directly with the model's prediction engine prevents data leakage and makes the code clean and ready for software deployment.
---
5. How this project connects to real-world HR automation  
This project perfectly mirrors the underlying architecture of modern Applicant Tracking Systems (ATS). In real-world HR departments, recruiters receive thousands of resumes for a single open role, making manual review impossible. An AI pipeline, like the one built in Task 13, automates the initial screening bottleneck. By instantly parsing resume text, weighing candidate metrics, and generating an "AI Confidence Score," this automation allows HR teams to focus their human interviews strictly on the top tier of candidates—drastically reducing time-to-hire.